#############################################################################
#### Gene expression analysis of histone deacetylase 1 (HDAC1) knockout mouse 
#### Affymetrix microarray
#### Dataset: GSE5583 (http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE5583)
#### Paper: Mol Cell Biol 2006 Nov;26(21):7913-28.
#### PMID: 16940178 (http://www.ncbi.nlm.nih.gov/pubmed/16940178)

##############################################################################




```{r}
library(RCurl)
url = getURL ("http://bit.ly/GSE5583_data", followlocation = TRUE)
data = as.matrix(read.table (text = url, row.names = 1, header = T))
```

```{r}
dim(data)
head(data)
tail(data)
```


#####################
Exploratory Plots
#####################

```{r}
hist(data, col="gray", main="GSE5583 - Histogram")

#log2 transformation
data2<- log2(data)

# check the behavior of the data after the log transformation
hist(data2, col="gray", main="GSE5583(log2) - Histogram")

```

```{r}
# Boxplot
boxplot(data2, col=c("darkgreen", "darkgreen", "darkgreen",
	"darkred", "darkred", "darkred"),
	main="GSE5583 - boxplots", las=2)
```

```{r}
# Hierarchical clustering of the "samples" based on the correlation coefficients of the expression values
hc = hclust(as.dist(1-cor(data2)))
plot(hc, main="GSE5583 - Hierarchical Clustering")
```

######################################
Differential expression (DE) analysis
######################################

# Separate the two conditions into two smaller data frames

```{r}
wt <- data2[,1:3] # all rows but only columns 1-3
ko <- data2[,4:6]
```

# Compute the means of the samples of each condition

```{r}
wt.mean<-apply(wt,1,mean)
ko.mean<-apply(ko,1,mean)

head(wt.mean)
head(ko.mean)

```

# Get the maxium of all the means 

```{r}
limit=max(wt.mean,ko.mean)
```

# Scatter Plot

```{r}
plot(ko.mean~wt.mean, xlab="WT", ylab="KO", main="GSE5583 - Scatter", xlim=c(0,limit), ylim=c(0,limit))

```

# Compute fold-change (biological significance)

```{r}
# Difference between the means of the conditions

fold=wt.mean-ko.mean

#Histogram of the fold differences
hist(fold, col="gray")
```

# Compute statistical significance of the dataset using t-test

```{r}
pvalue=NULL #empty list for p-values
tstat=NULL 

for(i in 1: nrow(data)) {# for each gene:
  x=wt[i,] #WT of gene number i
  y=ko[i,] #KO of gene number i
  
  #Compute t-test between the 2 conditions
  t=t.test(x,y)
  
  #Put the current p-value in the pvalues list
  pvalue[i]=t$p.value
  
  # Do the same with t-stat value
  tstat[i]=t$statistic
}
  
```

```{r}
head(pvalue)
```

# Histogram of p-values

```{r}
hist(-log10(pvalue), col="gray")
```

# Make a volcano plot: put the biological significance (fold-change) and statistical significance (p-value) in one plot

```{r}
plot(fold, -log10(pvalue), main="GSE5583 - Volcano")

#set the cut-offs
fold_cutoff=2
pvalue_cutoff=0.01

abline(v = fold_cutoff, col = "blue", lwd = 3) # put a blue vertical line at the fold cutoff value
abline(v = -fold_cutoff, col = "red", lwd = 3) # put a red vertical line at the negative fold cutoff value
abline(h = -log10(pvalue_cutoff), col = "green", lwd = 3) # put a green horizontal line at the pvalue cutoff
```

```{r}
# Screen for the genes that satisy the filtering criteria

# Fold change filter for "biological" significance
filter_by_fold=abs(fold) >= fold_cutoff #keep the ones where the fold value is greater than or equal to the fold cutoff value

dim(data2[filter_by_fold,])

# # P-value filter for "statistical" significance
filter_by_pvalue = pvalue <= pvalue_cutoff
dim(data2[filter_by_pvalue, ])
```

```{r}
# Combine filter (both biological and statistical)

filter_combined=filter_by_fold & filter_by_pvalue

filtered=data2[filter_combined,]

dim(filtered)
head(filtered)
```

# Generating another volcano plot highlighting the significantly expressed genes

```{r}
plot(fold, -log10(pvalue), main="GSE5583 - Volcano #2")
points (fold[filter_combined], -log10(pvalue[filter_combined]),
	pch = 16, col = "red")

# Highlighting the up-regulated and down-regulated genes in blue

plot(fold, -log10(pvalue), main = "GSE5583 - Volcano #3")

points (fold[filter_combined & fold < 0],
	-log10(pvalue[filter_combined & fold < 0]),
	pch = 16, col = "red")

points (fold[filter_combined & fold > 0],
	-log10(pvalue[filter_combined & fold > 0]),
	pch = 16, col = "blue")

```

```{r}
# Cluster the rows (genes) and columns(samples) by correlation
library(rlang)
library(ggdendro)
library(gplots)
library(pvclust)
rowv=as.dendrogram(hclust(as.dist(1-cor(t(filtered)))))
colv=as.dendrogram(hclust(as.dist(1-cor(filtered))))

#Generate a heatmap

heatmap(filtered, Rowv=rowv, Colv=colv, cexCol=0.7)
```

```{r}
library (gplots)

# Enhanced heatmap
heatmap.2(filtered, Rowv=rowv, Colv=colv, cexCol=0.7, col=rev(redblue(256)), scale="row")
```

# Saving the heat map

```{r}
# Save the heatmap to a PDF file
pdf ("GSE5583_DE_Heatmap.pdf")
heatmap.2(filtered, Rowv=rowv, Colv=colv, cexCol=0.7,
	col = rev(redblue(256)), scale = "row")
dev.off()

# Save the DE genes to a text file
write.table (filtered, "GSE5583_DE.txt", sep = "\t",
	quote = FALSE)
```

